fix: set default values for indexer that permit forward progress on indexing without requiring 14 archive nodes

galxy25 opened the pull request:
- document how to check the status of the blockscout indexer process
- document how to run explorer locally to index a production network (e.g. mainnet or public testnet)

initial testing shows that a single blockscout process running on my computer with the settings in `.env` are able to stay at tip AND index older blocks at a reasonable clip without overwhelming either a single node locally or a cluster of remote nodes

galxy25 commented on .env:
```diff
@@ -39,8 +39,27 @@ SECRET_KEY_BASE=56NtB48ear7+wMSf0IQuWDAAazhpb31qyc7GiyspBP2vh7t5zlCsF5QDv76chXeN
+INDEXER_BLOCK_REWARD_CONCURRENCY=1
+INDEXER_RECEIPTS_CONCURRENCY=1
+INDEXER_COIN_BALANCES_CONCURRENCY=1
+INDEXER_CATCHUP_BLOCK_INTERVAL=5s
```
obvs will set higher / faster values for prod where blockscout can run against more / beefier nodes than a single kava node locally

nddeluca commented on .env:
```diff
@@ -39,8 +39,27 @@ SECRET_KEY_BASE=56NtB48ear7+wMSf0IQuWDAAazhpb31qyc7GiyspBP2vh7t5zlCsF5QDv76chXeN
+INDEXER_COIN_BALANCES_CONCURRENCY=1
+INDEXER_CATCHUP_BLOCK_INTERVAL=5s
+# The block number, where import of blocks by catchup fetcher begins from.
+FIRST_BLOCK=0
```
This should be 1?

galxy25 replied:
will test and document the result as part of the config of the deployed services

galxy25 replied:
for local 0 is correct because the dockerized kava node has / makes that block

nddeluca replied:
There is no block 0 in tendermint, the genesis block is height 1 and one of the differences between ethereum and cosmos based blockchains.

You are correct that it responds, but it should error.  It looks like a request for height zero actually responds with height 1 in etheremint, so probably a bug.

```
curl -sX POST -H "content-type: application/json" --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["1", true],"id":1}' https://evm.data.kava.io
```

nddeluca commented:
Great documentation, thanks for adding the query examples.

Another thing to look at is the `select count(*) from pending_block_operations;`, which is the leading cause of blockscout slowing down our nodes on mainnet is my hypothesis.

```
select count(*) from pending_block_operations;
> 1612632

select * from pending_block_operations limit 1;
> block_hash | inserted_at | updated_at | fetch_internal_transactions
> 0x95c132c930f104232464ab278cfde649e74963c743c634277497e824c38a43fc | 2023-01-21 23:49:21.372883 | 2023-01-21 23:49:21.372883 | true
```

I think once we deploy also need to update the rows in this table for fetch_internal_transactions = false, or delete all rows in the table then batch set refetch on blocks to re-index?

The number pending block operations might be a good metric for a cloudwatch alarm since a block isn't fully indexed until the pending operation row is deleted.

One thing I'm not sure on is if the pending_block_operations row also accounts for balance fetching, etc and tells us if all indexer types have finished with the block?

nddeluca commented:
One item I'd like to test is identify a block where a contract was deployed and wasn't correctly identified due to the kava 12 getCode bug, and see if flipping refetch corrects the issue.  If so, might be worth doing for all blocks from tip backward to correct data

galxy25 commented:
> Great documentation, thanks for adding the query examples.
> 
> Another thing to look at is the `select count(*) from pending_block_operations;`, which is the leading cause of blockscout slowing down our nodes on mainnet is my hypothesis.
> 
> ```
> select count(*) from pending_block_operations;
> > 1612632
> 
> select * from pending_block_operations limit 1;
> > block_hash | inserted_at | updated_at | fetch_internal_transactions
> > 0x95c132c930f104232464ab278cfde649e74963c743c634277497e824c38a43fc | 2023-01-21 23:49:21.372883 | 2023-01-21 23:49:21.372883 | true
> ```
> 
> I think once we deploy also need to update the rows in this table for fetch_internal_transactions = false, or delete all rows in the table then batch set refetch on blocks to re-index?
> 
> The number pending block operations might be a good metric for a cloudwatch alarm since a block isn't fully indexed until the pending operation row is deleted.
> 
> One thing I'm not sure on is if the pending_block_operations row also accounts for balance fetching, etc and tells us if all indexer types have finished with the block?

cc @nddeluca updated docs with additional query, added to list of potential metrics to track pending block operations to https://kava-labs.atlassian.net/wiki/spaces/ENG/pages/1168769042/Kava+Blockscout

added ticket to explore the need to set refetch on blocks dependent on ` fetch_internal_transactions = false` behavior

https://app.shortcut.com/kava-labs/story/10017/spikie-re-indexing-for-fetch-internal-transactions-false

galxy25 commented:
> One item I'd like to test is identify a block where a contract was deployed and wasn't correctly identified due to the kava 12 getCode bug, and see if flipping refetch corrects the issue. If so, might be worth doing for all blocks from tip backward to correct data

@nddeluca can you add a ticket for this with links to any (if we have it) documentation / shortcut tickets from those times in the past where we encountered and were debugging the kava 12 getCode bug? Not clear to me if this needs to go as part of the deployment epic or another future one but leaning toward the latter.
